File: app/src/main/java/com/derelicthut/basecode/mvvmkotlin/StatusBarCompat.java
package com.derelicthut.basecode.mvvmkotlin;

import android.os.Build;
import android.view.DisplayCutout;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowInsets;

import com.blankj.utilcode.util.BarUtils;

public class StatusBarCompat {

    private static int statusBarHeight = BarUtils.getStatusBarHeight();

    public static void expendStatusBarHeight(View view){
        if (view != null) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P){
                view.post(new Runnable() {
                    @Override
                    public void run() {
                        WindowInsets windowInsets = view.getRootWindowInsets();
                        if (windowInsets!= null){
                            DisplayCutout displayCutout = windowInsets.getDisplayCutout();
                            if (displayCutout!= null){
                                statusBarHeight = Math.max(statusBarHeight,displayCutout.getSafeInsetTop());
                            }
                        }
                        realExpand(view);
                    }
                });
            }else {
                realExpand(view);
            }
        }
    }

    public static void realExpand(View view) {
        ViewGroup.LayoutParams params = view.getLayoutParams();
        if (params != null) {
            params.height = params.height + statusBarHeight;
            view.setPadding(view.getPaddingLeft(), view.getPaddingTop() + statusBarHeight
                    , view.getPaddingRight(), view.getPaddingBottom());
        }
    }
}

File: base/src/main/java/com/derelicthut/basecode/base/utils/StatusBarCompat.kt
package com.derelicthut.basecode.base.utils

import android.os.Build
import android.view.View
import com.blankj.utilcode.util.BarUtils

/**
 * @author wjl
 */
object StatusBarCompat {
    private var statusBarHeight = BarUtils.getStatusBarHeight()

    @JvmStatic
    fun expendStatusBarHeight(view: View?) {
        view?.apply {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
                post {
                    rootWindowInsets?.let { windowInsets ->
                        windowInsets.displayCutout?.let {
                            statusBarHeight = statusBarHeight.coerceAtLeast(it.safeInsetTop)
                        }
                    }
                    realExpand(view)
                }
            } else {
                realExpand(this)
            }
        }
    }

    @JvmStatic
    fun realExpand(view: View) {
        view.layoutParams?.apply {
            this.height = this.height + statusBarHeight
            view.setPadding(view.paddingLeft,
                view.paddingTop + statusBarHeight,
                view.paddingRight,
                view.paddingBottom)
        }
    }
}